--- tabix-0.2.6/perl/Makefile.PL.orig	2013-08-01 15:55:51.886385000 +0100
+++ tabix-0.2.6/perl/Makefile.PL	2013-08-01 15:56:04.891339000 +0100
@@ -2,7 +2,7 @@
 WriteMakefile(
 			  NAME         => 'Tabix',
 			  VERSION_FROM => 'Tabix.pm',
-			  LIBS         => ['-lz -L.. -ltabix'],
+			  LIBS         => ['-L.. -ltabix -lz'],
 			  DEFINE       => '-D_FILE_OFFSET_BITS=64 -D_USE_KNETFILE',
 			  INC          => '-I..',
 			 );
